python-oauth-web-app/atproto_oauth.py

```diff
@@ -0,0 +1,60 @@
+
+from urllib.parse import urlparse
+import requests
+
+def valid_authsrv_meta(obj, url):
+    fetch_url = urlparse(url)
+    issuer_url = urlparse(obj["issuer"])
+    assert issuer_url.hostname == fetch_url.hostname
+    assert issuer_url.scheme == "https"
+    assert issuer_url.port == None
+    assert issuer_url.path in ["", "/"]
+    assert issuer_url.params == ""
+    assert issuer_url.fragment == ""
+
+    assert "code" in obj["response_types_supported"]
+    assert "authorization_code" in obj["grant_types_supported"]
+    assert "refresh_token" in obj["grant_types_supported"]
+    assert "S256" in obj["code_challenge_methods_supported"]
+    assert "none" in obj["token_endpoint_auth_methods_supported"]
+    assert "private_key_jwt" in obj["token_endpoint_auth_methods_supported"]
+    assert "ES256" in obj["token_endpoint_auth_signing_alg_values_supported"]
+    #assert "refresh_token" in obj["scopes_supported"]
+    assert "profile" in obj["scopes_supported"]
+    assert "email" in obj["scopes_supported"] # TODO
+    if "subject_types_supported" in obj:
+        assert "public" in obj["subject_types_supported"]
+    assert True == obj["authorization_response_iss_parameter_supported"]
+    assert obj["pushed_authorization_request_endpoint"] is not None
+    assert True == obj["require_pushed_authorization_requests"]
+    assert "ES256" in obj["dpop_signing_alg_values_supported"]
+    if "require_request_uri_registration" in obj:
+        assert True == obj["require_request_uri_registration"]
+    assert True == obj["client_id_metadata_document_supported"]
+
+    return True
+
+def fetch_authsrv_meta(url):
+    # TODO: ensure URL is safe
+    # fetch auth server metadata
+    resp = requests.get(f"{url}/.well-known/oauth-authorization-server")
```

Comment:
This is bsky.social right? Any idea when this'll be available in prod? 😅

Reply:
for a self-hosted PDS, the "resource server" and "authorization server" are the same thing: just the PDS. independent folks don't need to talk to `bsky.social` for auth! that is important.

for the bluesky hosted PDSes, the "resource server" is the PDS instance itself (a "mushroom server" on the `*.bsky.network` domain, there are about 20 of them), and the "authorization server" is the "entryway" service running at `bsky.social`.

we pushed code to the PDS distro for self-hosted instances a week or two ago. we haven't enabled OAuth functionality on the bluesky production servers yet, though probably will sooner than later.

we are still working on dev docs/resources and more official comms on this as an "official announcement" when we start to really encourage devs to start adopting OAuth.